src/core/lib/lattice/hal/default/dcrtpoly.cpp

```diff
@@ -1740,6 +1740,62 @@ void DCRTPolyImpl<VecType>::ExpandCRTBasis(
     this->m_params = paramsQP;
 }
 
+template <typename VecType>
+void DCRTPolyImpl<VecType>::ExpandCRTBasisReverseOrder(
+    const std::shared_ptr<DCRTPolyImpl::Params> paramsQP, const std::shared_ptr<DCRTPolyImpl::Params> paramsP,
+    const std::vector<NativeInteger>& QHatInvModq, const std::vector<NativeInteger>& QHatInvModqPrecon,
+    const std::vector<std::vector<NativeInteger>>& QHatModp, const std::vector<std::vector<NativeInteger>>& alphaQModp,
+    const std::vector<DoubleNativeInt>& modpBarrettMu, const std::vector<double>& qInv, Format resultFormat) {
+    std::vector<PolyType> polyInNTT;
+
+    // if the input polynomial is in evaluation representation, store it for
+    // later use to reduce the number of NTTs
+    if (this->GetFormat() == Format::EVALUATION) {
+        polyInNTT = m_vectors;
+        this->SetFormat(Format::COEFFICIENT);
+    }
+
+    DCRTPolyType partP =
+        SwitchCRTBasis(paramsP, QHatInvModq, QHatInvModqPrecon, QHatModp, alphaQModp, modpBarrettMu, qInv);
```

Comment:
Uses exact (HPS) basis extension. Should we use approximate basis extension instead?

Reply:
@Nicholas Genise just talked to me about Exact versus Approx Basis Extension, and he pointed out that if the flooding error is b + u * Q' instead of b, then there could be some extra leakage, where an adversary can try to compute u, by modding the ciphertext by Q' and subtracting. The value of u could leak bits about what b is, so this probably reduces the level of security by some # of bits. Probably not worth trying to switch from Exact to Approx Basis Extension for a small performance improvement, if it reduces security.